### PR TITLE
refactor!: re-design the whole SDK according to the new capability-oriented plugin API

### PR DESCRIPTION
**Pull request by jasondellaluce**
**What type of PR is this?**

/kind cleanup

/kind design

**Any specific area of the project related to this PR?**

area build

/area plugin-sdk

**What this PR does / why we need it**:

The C++ SDK is totally unusable due to it being left behind the recent plugin API version updates. Plugins built on top of this SDK are not compatibile with the new versions of Falco and the plugin framework in general.

This PR proposes redesigning the SDK from scratch to match the principles we used in the SDK Go.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
refactor!: re-design the whole SDK according to the new capability-oriented plugin API
```

### CODE REVIEW

**Comment by jasondellaluce** [comment]
This is a big WIP and I'm just leaving the work here for early feedback and traceability. This will be rebased for a more clear git history, and will require in-depth testing and review before getting merged or released.

**Comment by jasondellaluce** [comment]
cc @FedeDP, @leogr

**Review by FedeDP — COMMENTED** [review]
Left some nitpicks; btw really nice work as always Jason!
I really like the new sdk!

File: examples/full/src/full.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+*/
+
+#pragma once
+
+#include <falcosecurity/sdk.h>
+
+class full_instance: public falcosecurity::event_sourcer::instance
+{
+public:
+    ss_plugin_rc next(const falcosecurity::event_sourcer* p, ss_plugin_event* evt);
```
[the remark this answers is not in the document]
Comment (jasondellaluce) [reply examples/full/src/full.h]: Totally, I lost track of this! Thanks ❤️

**Comment by FedeDP** [comment]
Imho this LGTM.
/cc @leogr

**Review by leogr — APPROVED** [review]
I can just say this is awesome 👍

Thank you!

**Pull request merged**